### PR TITLE
Improvement to use Siddhi streams to send notifications

### PR DESCRIPTION
**Pull request by IndunilRathnayake**

### CODE REVIEW

File: components/event-handler-email/org.wso2.carbon.identity.event.handler.email/src/main/java/org.wso2.carbon.identity.event.handler.email/util/NotificationUtil.java [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+    public static void deployPublisher(EventPublisherConfiguration eventPublisherConfiguration) throws NotificationEventRuntimeException {
+        EventPublisherService eventPublisherService = NotificationHandlerDataHolder.getInstance().getEventPublisherService();
+        try {
+            eventPublisherService.deployEventPublisherConfiguration(eventPublisherConfiguration);
+        } catch (EventPublisherConfigurationException e) {
+            throw NotificationEventRuntimeException.error("Error in deploying a publisher.", e);
+        }
+    }
+
+    public static Notification getNotification(Event event, Map<String, String> placeHolderData)
```
Comment (johannnallathamby) [inline components/event-handler-email/org.wso2.carbon.identity.event.handler.email/src/main/java/org.wso2.carbon.identity.event.handler.email/util/NotificationUtil.java]: Rename  to buildNotification()

File: components/event-handler-notification/org.wso2.carbon.identity.event.handler.notification/src/main/java/org.wso2.carbon.identity.event.handler.notification/internal/NotificationHandlerDataHolder.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+import org.wso2.carbon.event.stream.core.EventStreamService;
+import org.wso2.carbon.registry.core.service.RegistryService;
+import org.wso2.carbon.user.core.service.RealmService;
+import org.wso2.carbon.utils.ConfigurationContextService;
+import org.wso2.carbon.email.mgt.EmailTemplateManager;
+
+public class NotificationHandlerDataHolder {
+
+    private static volatile NotificationHandlerDataHolder instance = new NotificationHandlerDataHolder();
+    RealmService realmService = null;
```
Comment (GayanM) [inline components/event-handler-notification/org.wso2.carbon.identity.event.handler.notification/src/main/java/org.wso2.carbon.identity.event.handler.notification/internal/NotificationHandlerDataHolder.java]: put correct access modifiers

File: components/event-handler-notification/org.wso2.carbon.identity.event.handler.notification/src/main/java/org.wso2.carbon.identity.event.handler.notification/internal/NotificationHandlerServiceComponent.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+ * interface="org.wso2.carbon.email.mgt.EmailTemplateManager" cardinality="1..1"
+ * policy="dynamic" bind="setEmailTemplateManager" unbind="unsetEmailTemplateManager"
+ */
+
+public class NotificationHandlerServiceComponent {
+
+    private static Log log = LogFactory.getLog(NotificationHandlerServiceComponent.class);
+
+    protected void activate(ComponentContext context) {
+        context.getBundleContext().registerService(AbstractEventHandler.class.getName(), new NotificationHandler(), null);
```
Comment (GayanM) [inline components/event-handler-notification/org.wso2.carbon.identity.event.handler.notification/src/main/java/org.wso2.carbon.identity.event.handler.notification/internal/NotificationHandlerServiceComponent.java]: catch Exception during bundle register

File: components/event-handler-notification/org.wso2.carbon.identity.event.handler.notification/src/main/java/org.wso2.carbon.identity.event.handler.notification/email/bean/Notification.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        return this.body;
+    }
+
+    public String getFooter() {
+        return this.footer;
+    }
+
+    public static class EmailNotificationBuilder {
+
+        private String sendTo;
```
Comment (GayanM) [inline components/event-handler-notification/org.wso2.carbon.identity.event.handler.notification/src/main/java/org.wso2.carbon.identity.event.handler.notification/email/bean/Notification.java]: define Optional / Required parameters if there are any optional parameters assign them to default value

File: components/event-handler-notification/org.wso2.carbon.identity.event.handler.notification/src/main/java/org.wso2.carbon.identity.event.handler.notification/util/NotificationUtil.java [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+    private static Log log = LogFactory.getLog(NotificationUtil.class);
+
+    public static Map<String, String> getUserClaimValues(String userName, UserStoreManager userStoreManager) {
+
+        Claim[] userClaims;
+        Map<String, String> claimsMap = new HashMap<String, String>();
+        try {
+            userClaims = userStoreManager.getUserClaimValues(userName, UserCoreConstants.DEFAULT_PROFILE);
+            if (userClaims != null) {
+                for (int i = 0; i < userClaims.length; i++) {
```
Comment (GayanM) [inline components/event-handler-notification/org.wso2.carbon.identity.event.handler.notification/src/main/java/org.wso2.carbon.identity.event.handler.notification/util/NotificationUtil.java]: use for each

**Comment by IndunilRathnayake** [comment]
The approach taken is as follows.
1) In default pack there will be a stream and a publisher deployed in the deployement server. When a tenant get created the same stream and publisher definitions will be created and deployed into the server
2) When an email notification needs to be send, load the email template from the registry 
3) Extract the placeholders in email template
4) Get the user claims from userstore and fill out the placeholders in the template with the necessary claim vlaues
5) Publish an arbitrary map including the filled out template to the stream

**Pull request merged**